File: clojure/P07.clj
; P07 Flatten a nested list structure.

; Predefined functions
(defn f0 [l]
  (flatten l))

; Recursion
(defn f1 [l]
  (if (seq l)
    (let [h (first l) t (rest l)]
      (if (list? h)
        (concat (f1 h) (f1 t))
        (cons h (f1 t))))
    '()))

File: java/P07.java
// P07 Flatten a nested list structure.

import core.List;

import static core.List.*;

class P07 {
    // Recursion, unchecked
    @SuppressWarnings("unchecked")
    List f1(List list) {
        if (list.isEmpty()) return nil();
        Object h = list.head();
        List t = list.tail();
        return h instanceof List ? concat(f1((List) h), f1(t)) : cons(h, f1(t));
    }
}